pycqed/analysis/three_state_rotation.py

```diff
@@ -0,0 +1,146 @@
+    invA = matA.getI()
+    resultant = invA * matB
+    return resultant[0,0], resultant[1,0]
+
+
+def perpSlope(slope, eps=1e-30):
+    # takes slope(float) and returns the slope of a line perpendicular to it
+    if np.abs(slope) < eps:
+        return 1/eps
+    return (slope * -1) ** -1
+
+
+def lineFromSlope(slope, point):
+    """
+    slope is a float of slope
+    point is a tuple of ...
+    returns tuple of slope and y intercept
+    """
+    return slope, ((slope * (-1 * point[0])) + point[1])
+
+
+def find_heights_and_intersections(points, eps=1e-10):
+    """
+    Find heights of triangle and intersection with opposite side
+    """
+    sides = []
+    heights = []
+    height_vertices = []
+    intersections = []
+    if isinstance(points, np.ndarray):
+        points = points.tolist()
+    for p in it.combinations(points, 2):
+        side = getLine(p[0], p[1])
+        sides.append(side)
+        perp_slope = perpSlope(side[0])
+        h_v = [pt for pt in points if pt not in p][0]
+
+        height = lineFromSlope(slope=perp_slope, point=h_v)
+        heights.append(height)
+        height_vertices.append(h_v)
+    for h in heights:
+        temp = []
+        for s in sides:
+            try:
+                if np.abs(h[0]*s[0] + 1.) < eps:
+                    temp.append(getIntersection(h, s))
+                elif (np.abs(h[0]) < eps and np.abs(s[0]) > 1./eps) or \
+                        (np.abs(s[0]) < eps and np.abs(h[0]) > 1./eps):
+                    temp.append(getIntersection(h, s))
+            except np.linalg.LinAlgError:
+                print("No intersection / coincidence between height {} "
+                      "and side {}".format(h,s))
+
+        inter = temp
+        assert len(inter) == 1
+        intersections.append(inter[0])
+    return sides, heights, height_vertices, intersections
+
+
+def predict_pt_proba(pt, height_vertices, intersections, vertices):
+    proba = []
+    if isinstance(vertices, np.ndarray): # for ease of comparison
+        vertices = vertices.tolist()
+    for inter, h_v in zip(intersections, height_vertices):
+        side_vert = [v for v in vertices if v != h_v][0] - np.array(h_v)
+        inter_translated = np.array(inter) - np.array(h_v)
+        pt_translated = np.array(pt) - np.array(h_v)
+        inter_normalized = inter_translated/np.linalg.norm(
+            inter_translated, ord=2, keepdims=True)
+        proj = np.dot(pt_translated, inter_normalized) / \
+               np.dot(inter_normalized, side_vert)
+        proba.append(1 - proj)
+    return proba
+
+
+def predict_proba_avg_ro(X, cal_points):
+    # Possible improvements: make for any number of cal points and channels,
+    # remove (convenient) dependency on shapely
+    """
+    Predicts probabilities of each state of average readout data
+    X using calibration points for 3 (fixed number) of calibration points
+    Args:
+        X: (n_samples, n_channels) average readout data points.
```

Comment:
Docstring is outdated: there is no more "proj_pt_on_triangle".
I would to the docstring that probabilities returned can be negative and larger than 1. The mapping to the [0,1] interval would require the code that I had earlier, or a smarter way that does not depend on the shapely package. Do we think this is necessary/woth to concentrate on for the medium term? if yes, I could create a git issue for it (and let this code go as it is for now). @stephlazar  what do you think?

Reply:
I think this would definitely be nice to have since after this pull request, all our measurements with f-level will use this analysis.
Maybe it makes sense that you implement it in this pull request and we merge it only when that is done. What do you think?

Reply:
I would rather do the opposite, let's merge this one for now to move on and I will create a git issue to remember we have to tackle this issue but it's not top 1 on the priority list right now, I believe. But it would be great to be able to use the code in this pull request already in the meanwhile.

Reply:
see issue #75 .